File: java/src/PriorityQueue.java
package src;

import java.util.Comparator;
// import src.Node;

/*
A riority queue implemented using a singly linked list. 
The PriorityQueue class has two constructors: one that uses the natural ordering of the elements and another that accepts a custom comparator for priority ordering. 
The enqueue method inserts an item into the priority queue according to its priority, while the dequeue method removes and returns the highest priority item from the queue. 
The peek method returns the highest priority
*/

public class Node<T> {
    T value;
    Node<T> next;

    public Node(T value) {
        this.value = value;
        this.next = null;
    }
}

public class PriorityQueue<T extends Comparable<T>> {
    private int length;
    private Node<T> head;
    private Comparator<T> comparator;

    // Default constructor using natural ordering
    public PriorityQueue() {
        this.head = null;
        this.length = 0;
        this.comparator = null;
    }

    // Constructor with custom comparator for priority ordering
    public PriorityQueue(Comparator<T> comparator) {
        this.head = null;
        this.length = 0;
        this.comparator = comparator;
    }

    // Compare two items using the provided comparator or natural ordering
    private int compare(T item1, T item2) {
        if (comparator == null) {
            return ((Comparable<T>) item1).compareTo(item2);
        } else {
            return comparator.compare(item1, item2);
        }
    }

    public void enqueue(T item) {
        Node<T> newNode = new Node<>(item);

        // If the queue is empty or the new item has higher priority than the current head
        if (this.head == null || compare(item, this.head.value) > 0) {
            newNode.next = this.head;
            this.head = newNode;
        } else {
            Node<T> currentNode = this.head;

            // Traverse the list until we find the correct position for the new item
            while (currentNode.next != null && compare(item, currentNode.next.value) <= 0) {
                currentNode = currentNode.next;
            }

            // Insert the new node after the currentNode
            newNode.next = currentNode.next;
            currentNode.next = newNode;
        }

        this.length++;
    }

    public T dequeue() {
        if (this.head == null) {
            return null;
        }

        this.length--;

        Node<T> oldHead = this.head;
        this.head = this.head.next;

        oldHead.next = null;

        return oldHead.value;
    }

    public T peek() {
        return this.head != null ? this.head.value : null;
    }

    public int size() {
        return this.length;
    }


    public static void main(String[] args) {
        // Create a PriorityQueue using natural ordering
        PriorityQueue<Integer> queue = new PriorityQueue<>();

        // Enqueue some elements
        queue.enqueue(3);
        queue.enqueue(1);
        queue.enqueue(5);
        queue.enqueue(2);

        // Dequeue and print the elements in descending order
        while (queue.size() > 0) {
            System.out.println(queue.dequeue());
        }
    }
}

// USAGE:
// Create a PriorityQueue with a custom Comparator to determine priority ordering. 
// If no comparator is provided, the natural ordering will be used
// Comparator<Integer> customComparator = (a, b) -> b - a;
// PriorityQueue<Integer> priorityQueue = new PriorityQueue<>(customComparator);

File: rust/src/priority_queue.rs
use std::cmp::Ordering;
use std::rc::Rc;

pub struct Node<T> {
    value: T,
    next: Option<Box<Node<T>>>,
}

pub struct PriorityQueue<T, F>
where
    T: PartialOrd,
    F: Fn(&T, &T) -> Ordering,
{
    head: Option<Box<Node<T>>>,
    size: usize,
    comparator: Rc<F>,
}

impl<T, F> PriorityQueue<T, F>
where
    T: PartialOrd,
    F: Fn(&T, &T) -> Ordering,
{
    // Constructor with custom comparator for priority ordering
    pub fn new(comparator: F) -> Self {
        Self {
            head: None,
            size: 0,
            comparator: Rc::new(comparator),
        }
    }

    // Compare two items using the provided comparator
    fn compare(&self, item1: &T, item2: &T) -> Ordering {
        (self.comparator)(item1, item2)
    }

    // Enqueue an item into the priority queue
    pub fn enqueue(&mut self, item: T) {
        let new_node = Box::new(Node {
            value: item,
            next: None,
        });

        if self.head.is_none() || self.compare(&new_node.value, &self.head.as_ref().unwrap().value) == Ordering::Greater {
            new_node.next = self.head.take();
            self.head = Some(new_node);
        } else {
            let mut current_node = self.head.as_mut().unwrap();

            while current_node.next.is_some()
                && self.compare(&new_node.value, &current_node.next.as_ref().unwrap().value) != Ordering::Greater
            {
                current_node = current_node.next.as_mut().unwrap();
            }

            new_node.next = current_node.next.take();
            current_node.next = Some(new_node);
        }

        self.size += 1;
    }

    // Dequeue the highest priority item from the priority queue
    pub fn dequeue(&mut self) -> Option<T> {
        self.head.take().map(|old_head| {
            self.head = old_head.next;
            self.size -= 1;
            old_head.value
        })
    }

    // Peek at the highest priority item without removing it
    pub fn peek(&self) -> Option<&T> {
        self.head.as_ref().map(|head| &head.value)
    }

    // Get the size of the priority queue
    pub fn size(&self) -> usize {
        self.size
    }
}


// USAGE example with integers:
// In this Rust implementation, the PriorityQueue struct accepts a custom comparator closure with a reference-counted pointer Rc to avoid multiple ownership issues. If you want to use the natural order for a type implementing Ord, you can pass std::cmp::Ord::cmp as the comparator function.
// fn main() {
//     let custom_comparator = |a: &i32, b: &i32| b.cmp(a);
//     let mut priority_queue = PriorityQueue::new(custom_comparator);

//     priority_queue.enqueue(3);
//     priority_queue.enqueue(1);
//     priority_queue.enqueue(2);

//     println!("Dequeued: {:?}", priority_queue.dequeue()); // 3
//     println!("Peek: {:?}", priority_queue.peek()); // 2
//     println!("Size: {}", priority_queue.size()); // 2
// }
